crates/rpc/rpc/src/eth/filter.rs

```diff
@@ -57,17 +67,84 @@ where
-        todo!()
+    async fn filter_changes(&self, id: SubscriptionId<'_>) -> RpcResult<FilterChanges> {
+        let id = id.into_owned();
+        let info = self.inner.client.chain_info().to_rpc_result()?;
+        let best_number = info.best_number;
+
+        let (start_block, kind) = {
+            let mut filters = self.inner.active_filters.inner.lock().await;
+            let mut filter = filters.get_mut(&id).ok_or_else(|| FilterError::FilterNotFound(id))?;
+
+            // update filter
+            // we fetch all changes from [filter.block..best_block], so we advance the filter's
+            // block to `best_block +1`
+            let mut block = best_number + 1;
+            std::mem::swap(&mut filter.block, &mut block);
+            filter.last_poll_timestamp = Instant::now();
+
+            (block, filter.kind.clone())
+        };
+
+        match kind {
+            FilterKind::PendingTransaction => {
+                return Err(internal_rpc_err("method not implemented"))
+            }
+            FilterKind::Block => {
+                let mut block_hashes = Vec::new();
+                for block_num in start_block..best_number {
+                    let block_hash = self
+                        .inner
+                        .client
+                        .block_hash(U256::from(block_num))
+                        .to_rpc_result()?
+                        .ok_or(EthApiError::UnknownBlockNumber)?;
+                    block_hashes.push(block_hash);
+                }
+                Ok(FilterChanges::Hashes(block_hashes))
+            }
+            FilterKind::Log(filter) => {
+                let mut from_block_number = start_block;
+                let mut to_block_number = best_number;
+                match filter.block_option {
+                    FilterBlockOption::Range { from_block, to_block } => {
+                        // from block is maximum of block from last poll or `from_block` of filter
+                        if let Some(filter_from_block) =
+                            from_block.and_then(|num| info.convert_block_number(num))
+                        {
+                            from_block_number = start_block.max(filter_from_block)
+                        }
+
+                        // to block is max the best number
+                        if let Some(filter_to_block) =
+                            to_block.and_then(|num| info.convert_block_number(num))
+                        {
+                            to_block_number = filter_to_block;
+                            if to_block_number > best_number {
+                                to_block_number = best_number;
+                            }
+                        }
+                    }
+                    FilterBlockOption::AtBlockHash(_) => {
+                        // blockHash is equivalent to fromBlock = toBlock = the block number with
+                        // hash blockHash
```

Comment:
yes the range by default is the entire range and the blockHash is basically

> The Ethereum node responding to the request would either send back an error if the block hash was not found or it would return the results matching the filter (per normal operation) constrained to the block provided. Internally, this would function (presumably) similar to the fromBlock and toBlock filter options.

https://eips.ethereum.org/EIPS/eip-234

so we need to loop over the range and find matching block hash I suppose, but not entirely sure yet.

Reply:
let's track in new ticket perhaps?